python_ta/typecheck/base.py

```diff
@@ -219,105 +261,104 @@ def _find(self, tv: TypeVar) -> _TNode:
                 rep1.parent = self._make_set(t2)
-                return t2
+                return self.resolve(t1)
             else:
                 return self.unify(rep1.type, t2)
         elif isinstance(t2, TypeVar):
             return self.unify(t2, t1)
-        elif isinstance(t1, GenericMeta) and isinstance(t2, GenericMeta):
-            return self._unify_generic(t1, t2)
-        elif isinstance(t1, GenericMeta) or isinstance(t2, GenericMeta):
-            return f'Incompatible types {t1} {t2}'
-        elif t1.__class__.__name__ == '_Union' or t2.__class__.__name__ == '_Union':
-            return t1
-        elif t1 == Any or t2 == Any:
-            return t1
-        elif isinstance(t1, _ForwardRef) and isinstance(t2, _ForwardRef) and t1 == t2:
-            return t1
-        elif isinstance(t1, _ForwardRef) or isinstance(t2, _ForwardRef):
-            return f'Incompatible types {t1} {t2}'
+        elif isinstance(t1, _ForwardRef) and \
+             isinstance(t2, _ForwardRef) and t1 == t2:
+            return TypeInfo(t1)
+
+        # Case of unifying two concrete types
         elif t1 == t2:
-            return t1
-        elif t1 != t2:
-            return f'Incompatible types {t1} {t2}'
-
-    def _unify_generic(self, t1: GenericMeta, t2: GenericMeta) -> Union[str, GenericMeta]:
-        """Unify two generic types (e.g., List, Tuple, Dict, Callable)."""
-        if _gorg(t1) is not _gorg(t2):
-            return f'Incompatible generic types {_gorg(t1)} and {_gorg(t2)}'
-        elif t1.__args__ is not None and t2.__args__ is not None:
-            # TODO: check for equal lengths of __args__
-            gorg = _gorg(t1)
-            if gorg == Callable:
-                # TODO: Looking up elements in arrays causes _type_check to be
-                # called from typing.py, which tries to create a ForwardRef
-                # when encountering strings, even though strings are sometimes
-                # error messages
-                return gorg[
-                    [self.unify(a1, a2) for a1, a2 in zip(t1.__args__[:-1], t2.__args__[:-1])],
-                    self.unify(t1.__args__[-1], t2.__args__[-1])
-                ]
-            elif gorg == Tuple:
-                tuple_args = tuple(
-                    self.unify(a1, a2) for a1, a2 in zip(t1.__args__, t2.__args__)
-                )
-                # Handle the special case when t1 or t2 are empty tuples
-                if tuple_args == ((),):
-                    tuple_args = ()
-                return gorg[tuple_args]
-            else:
-                return gorg[tuple(
-                    self.unify(a1, a2) for a1, a2 in zip(t1.__args__, t2.__args__)
-                )]
+            return TypeInfo(t1)
         else:
-            return f'Incompatible types {t1} and {t2}'
+            return TypeFail(f'Incompatible types {t1} and {t2}')
+
+    def _unify_generic(self, t1: GenericMeta, t2: GenericMeta) -> TypeResult:
```

Comment:
You have changed the logic here too much. The *only* change that should have been made is the return type.